drivers/i2c/i2c_ll_stm32.c

```diff
@@ -283,9 +358,48 @@ static int i2c_stm32_init(const struct device *dev)
 		return ret;
 	}
 
+#if IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)
+	i2c_stm32_suspend(dev);
+	pm_device_init_suspended(dev);
+	(void)pm_device_runtime_enable(dev);
+#endif
+
 	return 0;
 }
 
+#ifdef CONFIG_PM_DEVICE
+
+static int i2c_stm32_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	int err;
+	const struct i2c_stm32_config *cfg = dev->config;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		err = i2c_stm32_activate(dev);
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		/* Do not suspend if I2C is enabled because it means the driver is waiting for
+		 * an interrupt or a timeout expiration. We can conveniently check the
+		 * I2C state here because the driver disables I2C as soon as it can
+		 * free the peripheral.
+		 * If PM_DEVICE_RUNTIME is enabled, suspend action shouldn't be used until allowed
+		 * in any case.
+		 */
+		if (LL_I2C_IsEnabled(cfg->i2c) != 0) {
+			return -EBUSY;
+		}
```

Comment:
According to implementation of function `pm_suspend_devices` which will trigger this code, answering `-EBUSY` will end up in a different behavior than if you had used `pm_device_busy_set` or `pm_policy_state_lock_put` in your driver.
See `pm_suspend_devices` implementation: receiving `-EBUSY` will end up returning an error and `pm_resume_devices` will be run, waking up all devices that were previously suspended

Reply:
The PM API is a bit more complex than I thought and I am wondering if I should keep things simple here by only using `pm_device_busy_set` and by removing support for `PM_DEVICE_RUNTIME` at this point. 

What do you think?

Reply:
I finally rebased to use `pm_device_runtime_get`/`pm_device_runtime_put` when `CONFIG_PM_DEVICE_RUNTIME` is enabled and `pm_device_busy_set`/`pm_device_busy_clear` if not